packages/mui-base/src/utils/useAnimationUnmount.ts

```diff
@@ -0,0 +1,55 @@
+'use client';
+import * as React from 'react';
+import { useEventCallback } from './useEventCallback';
+import { ownerWindow } from './owner';
+
+/**
+ * Unmounts the supplied element only if it has no animations or transitions.
+ * @ignore - internal hook.
+ */
+export function useAnimationUnmount(
+  getElement: () => Element | null | undefined,
+  onUnmount: () => void,
+) {
+  const frame1Ref = React.useRef(-1);
+  const frame2Ref = React.useRef(-1);
+
+  const cancelFrames = useEventCallback(() => {
+    cancelAnimationFrame(frame1Ref.current);
+    cancelAnimationFrame(frame2Ref.current);
+  });
+
+  React.useEffect(() => cancelFrames, [cancelFrames]);
+
+  const unmount = useEventCallback(() => {
+    cancelFrames();
+
+    const element = getElement();
+
+    if (!element) {
+      return;
+    }
+
+    // Wait for the CSS styles to be applied to determine if the animation has been removed in the
+    // [data-instant] state. This allows the close animation to play if the `delay` instantType is
+    // applying to the same element.
+    // Notes:
+    // - A single `requestAnimationFrame` is sometimes unreliable.
+    // - `queueMicrotask` does not work.
+    frame1Ref.current = requestAnimationFrame(() => {
```

Comment:
I'm wondering if two animation frames are guaranteed to be reliable.
If we detect an animation where there isn't one (the `[data-instant]` example), we may end up with an element waiting indefinitely to be unmounted, breaking the user interface.

Reply:
Playing around with the transitions experiments demo with `[data-instant]`, I wasn't ever able to reproduce it not unmounting. If there isn't any animation at all (no `[data-instant]` or anything), this won't be an issue at all.